### PR TITLE
[MPS] prod reduction: outer/inner shape specializations (perf)

### PR DESCRIPTION
**Pull request by anagnorisis2peripeteia**
Adds shape-specialized Metal prod kernels on top of the native generic
prod_reduction kernel from #188046 (the core migration). A single-dim
contiguous reduction is routed by shape to a specialization -- reduce over
dim 0 (outer), reduce over the last dim (inner), few rows of a huge reduced
dim (inner_wide), tall-thin small reduced dim (inner_thin), small outer M
(outer_smallm) -- and falls back to the generic kernel otherwise. The
specializations carry M and N as uint32, so a reduction whose dimension or
non-reduced product exceeds 2^32 falls to the guarded generic path. The host
grid (ceil_div + threadgroup-count multiply) and the outer kernel's per-y row
range are computed in 64-bit, so M or N immediately below 2^32 cannot wrap a
uint32 to zero and silently dispatch nothing / skip the reduction loop.

## Context
Part of #187455. Second piece of splitting #187788 (the ~1600-line native
prod migration) into reviewable PRs at the reviewer's request; stacks on
#188046 (prod-core). #188046 is correctness-focused (parity with the MPSGraph
reduction + removes prod from the shape-dependent MPSGraph cache); this PR
adds the steady-state speedups from the shape specializations.

## Perf (M4 Pro, single-dim contiguous prod, torch.utils.benchmark median,
vs the generic kernel from #188046)
  (4096, 4096) dim=0  (outer):       1032 us -> 428 us   (2.4x)
  (1000000, 8) dim=1  (inner_thin):  3215 us -> 358 us   (9.0x)
  (2, 8000000) dim=1  (inner_wide):  1536 us -> 1058 us  (1.45x)
  (4096, 4096) dim=1, (1024, 16384) dim=1, all-reduce: parity. The steady-state
  bench's per-iter synchronize floors fast shapes at ~400 us, so the specialized
  inner kernel reads as parity here; it is never a regression.
No regressions. Bench script: benchmarks/operator_benchmark/mps/prod_reduce_bench.py
(run on a build with and without this PR and compare).

## Test Plan
```
python test/test_mps.py -k test_prod_metal -v
```
19 tests, including test_prod_metal_wide_inner_few_rows and
test_prod_metal_thin_inner_small_n which exercise the inner_wide / inner_thin
specializations against the CPU reference. The existing outer/inner shapes in
the suite now route through the specialized kernels. (The >2^32 grid-overflow
path is not unit-tested: it needs a real ~17GB contiguous allocation, the same
constraint as the prod-core uint32 guard.)

## Known limitation (shared with sum_reduction / value_reduction)
The specialization kernels do not support a reduced extent or non-reduced product
within a threadgroup stride of 2^32 (reachable only via a crafted ~17GB contiguous
tensor). Two coupled uint32 overflows exist in that regime: the reduction loops use
uint32 induction (`row += NCHAINS`, `base += stride`, `i += tptg`), which wraps below
the loop bound and never terminates; and the outer kernel's flat index computes
`row + j` in uint32 before the ulong cast, so it too wraps near 2^32. Both patterns
mirror the existing sum_reduction / value_reduction specialization kernels (prod
already 64-bit-casts the row*N offset multiply, hardening it beyond the siblings,
but the inner `row + j` still wraps pre-cast), so the limitation is shared; the
remaining fix -- 64-bit loop induction and casting `row` before the `+ j` -- is
most consistent applied across all the reduce kernels in a follow-up rather than
diverging in prod alone. The host grid math is computed in
64-bit, so the oversized case does not silently no-op via an empty dispatch; because
the loop never terminates the net effect is a GPU hang, not a wrong-but-returned
result.

Review history (local-gate findings + dispositions): https://gist.github.com/anagnorisis2peripeteia/84aacf7575b484ab85bce4d5e1da165a

Authored with Claude.

### CODE REVIEW

**Comment by anagnorisis2peripeteia** [comment]
**Stacked on #188046 — the real size is much smaller than the diff shows.**

This builds on #188046 (the prod-core kernel) and is opened against `main` while that lands, so GitHub currently shows the **cumulative** diff (+1165 / 5 files, both commits). The size of *this* PR alone -- the shape specializations + dispatch, i.e. the top commit -- is **+581 / 4 files**, under the ~600-line target. Once #188046 merges, the diff here reduces to exactly that.

**Review the top commit** (`[MPS] prod reduction: outer/inner shape specializations`); the bottom commit is #188046.

Splits #187788 (closed), tracked by #187455. Local review-history (findings + dispositions): https://gist.github.com/84aacf7575b484ab85bce4d5e1da165a

**Comment by anagnorisis2peripeteia** [comment]
Closing in favor of #188156. The reduce-family consolidation in #188156 routes sum / mean / nansum / count_nonzero through a shared two-pass `value_reduction<Op, Load>` Metal kernel -- the same kernel min/max/all/any already use. Prod will be redone on top of #188156 as `value_reduction<ProdOp>` (adding a ProdOp functor) rather than a separate prod_reduction kernel plus its shape specializations, which removes the duplication flagged across the reduce PRs. The outer/inner specialization analysis here carries into that follow-up -- superseded, not abandoned.
